### DISCUSSION TITLE
On downgrade from 5.8.1 to 5.2.8, workers do not connect with the concourse web node

### DISCUSSION THREAD

**Discussion by prateeksisodia** [concourse/concourse#6198] [Help & Support]
We recently upgraded one of our concourse from 5.2.8 to 5.8.1, we have to downgrade it back to version 5.2.8 and on downgrade we are facing issue with workers connection to servers node.
We then tried to downgrade from 5.8.1 to 5.5.11 and in this case workers are able to connect but none of pipelines works due to below error:
`runc run: exit status 1: container_linux.go:349: starting container process caused "process_linux.go:439: container init caused \"write sysctl key net.ipv4.tcp_keepalive_probes: open /proc/sys/net/ipv4/tcp_keepalive_probes: no such file or directory\""`

We have to downgrade from 5.8.1 to 5.2.8, can someone please assist.
We tried the steps to downgrade as mentioned on following page and it did not work:
https://concourse-ci.org/concourse-web.html#downgrading

**Comment by jamieklassen** [comment]
Hi @prateeksisodia this isn't well-documented, but 5.5.x was an "LTS" version which the core Concourse contributors kept patching until July 31, 2020 (hence the 11 patches, compared to most minors which usually have 0 or 1). By "downgrading" to 5.5.11, you actually ended up "upgrading" the bundled `gdn`, and now you're hitting this breaking change: https://github.com/concourse/concourse/releases/tag/v5.5.11#gdn-breaking. I have discussed this in more detail here: https://github.com/concourse/concourse/issues/5711#issuecomment-638224176.

Sorry you wound up in this mess, I'm sure you wouldn't think to consider a "downgrade" as a "breaking change" in this way. You can most likely fix your _current_ problem by upgrading your kernel, or else maybe you can share details about why your initial 5.2.8 -> 5.8.1 upgrade failed?

Out of curiosity, are you deploying your workers via Docker on Ubuntu 16.04 by any chance? concourse/concourse-docker#61

**Reply by prateeksisodia** [reply]
Hi  @jamieklassen,
The upgrade didn't failed, but we have to downgrade to 5.2.8 as "DISABLE_LIST_ALL_JOBS" was not available in the v5.8.1 and also before upgrading other environments we had to ensure that a rollback is possible to the current version in case of any unidentified issues with the version 5.8.1.

Downgrade to 5.2.8 from 5.8.1 also did not give any error, but on worker instances we can see below error in logs
`concourse-server[14693]: {"timestamp":"2020-10-08T13:35:46.289434369Z","level":"error","source":"atc","message":"atc.register-worker.failed-to-save-worker","data":{"error":"pq: date/time field value out of range: \"1602162541\"","session":"41825"}}`

We are using ubuntu1604 for our worker instances.

**Reply by jamieklassen** [reply]
I am guessing that the problem with your 5.8.1 -> 5.2.8 downgrade is that **you did not down-migrate your database**. I just tried the following procedure:
1. **start with concourse 5.8.1** - replace `concourse/concourse` with `concourse/concourse:5.8.1` in https://github.com/concourse/concourse-docker/blob/master/docker-compose.yml and `docker-compose up -d`
1. **downtime** - `docker-compose stop web`
1. **down-migrate DB** - `docker run -e CONCOURSE_POSTGRES_HOST=db -e CONCOURSE_POSTGRES_USER=concourse_user -e CONCOURSE_POSTGRES_PASSWORD=concourse_pass -e CONCOURSE_POSTGRES_DATABASE=concourse --network concourse-docker_default -it concourse/concourse:5.8.1 migrate --migrate-db-to-version=1556724983`
1. **run concourse 5.2.8** - replace `concourse/concourse` with `concourse/concourse:5.2.8` in https://github.com/concourse/concourse-docker/blob/master/docker-compose.yml and `docker-compose up -d`

and I saw no errors. If instead, I run only these steps:

1. **start with concourse 5.8.1** - replace `concourse/concourse` with `concourse/concourse:5.8.1` in https://github.com/concourse/concourse-docker/blob/master/docker-compose.yml and `docker-compose up -d`
1. **run concourse 5.2.8** - replace `concourse/concourse` with `concourse/concourse:5.2.8` in https://github.com/concourse/concourse-docker/blob/master/docker-compose.yml and `docker-compose up -d`

my web logs are filled with errors just like yours:

```
web_1     | {"timestamp":"2020-10-23T18:14:51.495726800Z","level":"error","source":"atc","message":"atc.register-worker.failed-to-save-worker","data":{"error":"pq: date/time field value out of range: \"1603476854\"","session":"264"}}
```

**Reply by jamieklassen** [reply]
Reading between the lines a bit, I think you meant to say "web instances" instead of "worker instances"? Since the log you showed me is from the `atc` component? Guessing based on the format of those logs (`concourse-server[14693]:`) you're probably managing your concourse process with systemd? Maybe you're running everything in one VM?

**Reply by jamieklassen** [reply]
Also if you want to run a 5.5.11 worker process inside a user namespace (which I think systemd does) you'll need Linux kernel v4.5+, which in your case is maybe most readily achieved by upgrading to Ubuntu 18.04?

**Reply by chenbh** [reply]
It turns out Ubuntu 16.04.5 LTS has kernel version 4.15 and has the same support window as 16.04.0/1. That might fix the `net.ipv4.tcp_keepalive_probes` problem if upgrading to 18.04 is not possible.

https://ubuntu.com/kernel/lifecycle#support-16-04 (scroll down to the Kernel Support Schedule)
